base-trezor-keyring.js

```diff
@@ -0,0 +1,562 @@
+ */
+function isOldStyleEthereumjsTx(tx) {
+  return typeof tx.getChainId === 'function';
+}
+
+class BaseTrezorKeyring extends EventEmitter {
+  constructor(opts = {}) {
+    super();
+    this.type = keyringType;
+    this.accounts = [];
+    this.hdk = new HDKey();
+    this.page = 0;
+    this.perPage = 5;
+    this.unlockedAccount = 0;
+    this.paths = {};
+    this.deserialize(opts);
+  }
+
+  /**
+   * Performs any asynchronous initialisation
+   *
+   * @returns {Promise<void>}
+   */
+  init() {
+    throw new Error('Method init not implemented');
+  }
+
+  /**
+   * Gets the model, if known.
+   * This may be `undefined` if the model hasn't been loaded yet.
+   *
+   * @returns {"T" | "1" | undefined}
+   */
+  getModel() {
+    return this.model;
+  }
+
+  /**
+   * Clears the iframe created by Trezor
+   *
+   * @returns {Promise<void>}
+   */
+  dispose() {
+    throw new Error('Method dispose not implemented');
+  }
+
+  /**
+   * Gets public key from device
+   *
+   * @returns {Promise}
+   */
+  _getPublicKey(_payload) {
+    throw new Error('Method _getPublicKey not implemented');
+  }
+
+  /**
+   * Signs transaction using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignTransaction(_payload) {
+    throw new Error('Method _ethereumSignTransaction not implemented');
+  }
+
+  /**
+   * Signs message using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignMessage(_payload) {
+    throw new Error('Method _ethereumSignMessage not implemented');
+  }
+
+  /**
+   * Signs typed data using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignTypedData(_payload) {
+    throw new Error('Method _ethereumSignTypedData not implemented');
+  }
+
+  serialize() {
+    return Promise.resolve({
+      hdPath: this.hdPath,
+      accounts: this.accounts,
+      page: this.page,
+      paths: this.paths,
+      perPage: this.perPage,
+      unlockedAccount: this.unlockedAccount,
+    });
+  }
+
+  deserialize(opts = {}) {
+    this.hdPath = opts.hdPath || hdPathString;
+    this.accounts = opts.accounts || [];
+    this.page = opts.page || 0;
+    this.perPage = opts.perPage || 5;
+    return Promise.resolve();
+  }
+
+  isUnlocked() {
+    return Boolean(this.hdk && this.hdk.publicKey);
+  }
+
+  unlock() {
+    if (this.isUnlocked()) {
+      return Promise.resolve('already unlocked');
+    }
+    return new Promise((resolve, reject) => {
+      this._getPublicKey({
+        path: this.hdPath,
+        coin: 'ETH',
+      })
+        .then((response) => {
+          if (response.success) {
+            this.hdk.publicKey = Buffer.from(response.payload.publicKey, 'hex');
+            this.hdk.chainCode = Buffer.from(response.payload.chainCode, 'hex');
+            resolve('just unlocked');
+          } else {
+            reject(
+              new Error(
+                (response.payload && response.payload.error) || 'Unknown error',
```

Comment:
I tried using optional chaining, but the existing eslint config sets `ecmaVersion` to 2018. I believe we need 2020 for that. I'm happy to change it and use optional chaining if that doesn't cause any unwanted side effect.

Reply:
Ah, https://github.com/MetaMask/eth-trezor-keyring/blob/440074fce83de77f608b759ba07eff54619abe9e/package.json#L65 so 2020 should be good?

Reply:
This library is used in the extension, which still supports browsers that do not support optional chaining, and as a general rule we try to avoid introducing the need to transpile deps. We're working on dropping support for older browsers, and we've been migrating everything to TypeScript so that we can write using modern syntax. But for pure JS libraries it's better not to use newer syntax for now.